Estimate gas price

zachzundel opened the pull request:
So far, I have the gas price being queried from the API if the default is being used. I still need to fix the tests and maybe add a `--gasPrice` flag to the CLI. This is for issue #56 .

zachzundel commented:
@spalladino Could you take a look at this and let me know if it seems reasonable so far?

spalladino commented:
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

spalladino commented:
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

spalladino commented:
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object, to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still WIP).
- And a minor comment: these lines are repeated several times, see if it's possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

spalladino commented:
@zachzundel looks good so far! A few comments:

- Try to cache the gas price obtained from the API in the `state` object,
to avoid having to wait one query per each tx.
- Double-check the units in which the API returns the gas price, you may
need to add a few zeroes to the value you obtain.
- Make sure to add a few tests later down the road (I know this is still
WIP).
- And a minor comment: these lines are repeated several times, see if it's
possible to refactor them:
```js
   let fixedGasPrice = checkGasPrice(txParams)
   if (fixedGasPrice) {
      txParams.gasPrice = fixedGasPrice
  }
```

(sorry if you see this comment several times, I'm getting an error when
posting it)

On Fri, Oct 19, 2018 at 5:02 AM Zach Zundel <notifications@github.com>
wrote:

> @spalladino <https://github.com/spalladino> Could you take a look at this
> and let me know if it seems reasonable so far?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/zeppelinos/zos/pull/299#issuecomment-431230144>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAaOJN3NCC9v3a1vSW2eTMcvKbeoElw1ks5umUDSgaJpZM4Xt6TQ>
> .
>

zachzundel commented:
Hi @spalladino -- I've updated my PR and I think I've fixed everything you pointed out. There was already a test for estimateGas that seems to cover my use case, please let me know if that's not correct and I will write another one!

zachzundel commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,11 +156,28 @@ async function getNodeVersion () {
+    try {
+      let apiResponse = await axios.get(GAS_API_URL);
+      let gasPriceGwei = apiResponse.safeLow / 10;
+      let gasPrice = BN(gasPriceGwei * 1000000000)
```
Can do! Out of curiousity, can you point me to a writeup on why one should prefer const over let?

spalladino replied:
Sure! From [ESLint](https://eslint.org/docs/rules/prefer-const):

> const declaration tells readers, “this variable is never reassigned,” reducing cognitive load and improving maintainability.

zachzundel commented:
How can we tell testnet vs. mainnet here? Is it something in `txParams`? Do I need to use Truffle?

zachzundel commented:
@spalladino I've added in tests, this is ready for review!

spalladino commented on package.json:
```diff
@@ -3,5 +3,8 @@
     "lerna": "~3.3.2"
+  },
+  "dependencies": {
+    "8": "0.0.1"
```
What is this dependency for? Please remove if unneeded.

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
-  if (TRUFFLE_DEFAULT_GAS_PRICE.eq(gasPrice) || !gasPrice) {
-    throw new Error(`Cowardly refusing to execute transaction with excessively high default gas price of 100 gwei. Consider explicitly setting a different gasPrice value in your truffle.js config file. You can check reasonable values for gas price in https://ethgasstation.info/.`);
+
+  if (TRUFFLE_DEFAULT_GAS_PRICE == gasPrice || !gasPrice) {
```
Let's use bignumber equality functions for this, ie `TRUFFLE_DEFAULT_GAS_PRICE.eq(gasPrice)`

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
-    throw new Error(`Cowardly refusing to execute transaction with excessively high default gas price of 100 gwei. Consider explicitly setting a different gasPrice value in your truffle.js config file. You can check reasonable values for gas price in https://ethgasstation.info/.`);
+
+  if (TRUFFLE_DEFAULT_GAS_PRICE == gasPrice || !gasPrice) {
+    if (network != '1') {
```
We shouldn't need to do anything if network id is not mainnet, I think we can just leave gasPrice empty and let truffle pick the default. We can just `return` if `network != 1`.

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
+
+    txParams.gasPrice = await getETHGasStationPrice()
+
+    if (txParams.gasPrice >= TRUFFLE_DEFAULT_GAS_PRICE) {
```
Same here, let's use `TRUFFLE_DEFAULT_GAS_PRICE.lte(txParams.gasPrice)`

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -32,7 +36,7 @@ const TRUFFLE_DEFAULT_GAS_PRICE = BN(100000000000);
  */
 export async function sendTransaction(contractFn, args = [], txParams = {}, retries = RETRY_COUNT) {
-  await checkGasPrice(txParams);
+  txParams = await fixGasPrice(txParams)
```
`fixGasPrice` is both modifying `txParams` and returning it, which may be confusing to users of the function. Let's either: 1- remove the return value, and just call it as `await fixGasPrice(txParams)` (without the need for assigning to txParams)
2- or have fixGasPrice return a copy of txParams with gasPrice set (without modifying the parameter, so you'd need to make a copy of the txParams parameter among the first things you do in fixGasPrice)

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
+    txParams.gasPrice = await getETHGasStationPrice()
+
+    if (txParams.gasPrice >= TRUFFLE_DEFAULT_GAS_PRICE) {
+        throw new Error(`Gas price API gave very high value (>100gwei), please manually provide a gas price.`)
```
Let's rephrase to "The current gas price estimate from ethgasstation.info is over 100 gwei. If you do want to send a transaction with a gas price this high, please set it manually in your truffle.js configuration file."

spalladino commented on packages/lib/src/utils/Transactions.js:
```diff
@@ -152,12 +156,48 @@ async function getNodeVersion () {
+}
+
+async function getETHGasStationPrice() {
+    if (state.gasPrice) return state.gasPrice;
```
Please use two spaces for indentation, for consistency

spalladino commented on packages/lib/test/src/utils/Transactions.test.js:
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
     });
 
+    it('uses gas price API when gas not specified', async function () {
+        const oldNetwork = state.network
```
Please use two spaces for indentation, for consistency

spalladino commented on packages/lib/test/src/utils/Transactions.test.js:
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
+        const oldNetwork = state.network
+
+        state.network = "1";
+        const stub = sinon.stub(axios, 'get').resolves({ average: 49 })
```
Let's move these two lines to a beforeEach block. If needed, we can wrap these two tests in a "describe" block, so the beforeEach only affects them.

spalladino commented on packages/lib/test/src/utils/Transactions.test.js:
```diff
@@ -67,6 +73,34 @@ contract('Transactions', function([_account1, account2]) {
+        assertGasPrice(tx, 49 * 1e8);
+
+        stub.restore()
+        state.network = oldNetwork
```
Same here but with an afterEach

zachzundel commented:
I've pushed commits to address all your changes!
